libp2p-host/host.go

```diff
@@ -22,26 +31,88 @@ func main() {
 
 	// Set your own keypair
-	priv, _, err := crypto.GenerateEd25519Key(rand.Reader)
+	priv, _, err := crypto.GenerateKeyPair(
+		crypto.Ed25519, // Select your key type. Ed25519 are nice short
+		-1,             // Select key length when possible (i.e. RSA).
+	)
 	if err != nil {
 		panic(err)
 	}
 
+	var idht *dht.IpfsDHT
+
 	h2, err := libp2p.New(ctx,
-		// Use your own created keypair
+		// Use the keypair we generated
 		libp2p.Identity(priv),
-
-		// Set your own listen address
-		// The config takes an array of addresses, specify as many as you want.
-		libp2p.ListenAddrStrings("/ip4/0.0.0.0/tcp/9000"),
+		// Multiple listen addresses
+		libp2p.ListenAddrStrings(
+			"/ip4/0.0.0.0/tcp/9000",      // regular tcp connections
+			"/ip4/0.0.0.0/udp/9000/quic", // a UDP endpoint for the QUIC transport
+		),
+		// support TLS connections
+		libp2p.Security(libp2ptls.ID, libp2ptls.New),
+		// support secio connections
+		libp2p.Security(secio.ID, secio.New),
+		// support QUIC
+		libp2p.Transport(libp2pquic.NewTransport),
+		// support any other default transports (TCP)
+		libp2p.DefaultTransports,
+		// Let's prevent our peer from having too many
+		// connections by attaching a connection manager.
+		libp2p.ConnectionManager(connmgr.NewConnManager(
+			100,         // Lowwater
+			400,         // HighWater,
+			time.Minute, // GracePeriod
+		)),
+		// Attempt to open ports using uPNP for NATed hosts.
+		libp2p.NATPortMap(),
+		// Let this host use the DHT to find other hosts
+		libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
+			idht, err = dht.New(ctx, h)
+			return idht, err
+		}),
+		// Let this host use relays and advertise itself on relays if
+		// it finds it is behind NAT. Use libp2p.Relay(options...) to
+		// enable active relays and more.
+		libp2p.EnableAutoRelay(),
 	)
 	if err != nil {
 		panic(err)
 	}
 
+	// Now let's setup AutoNAT which helps figuring out if we are behind NAT
+	_, err = autonat.NewAutoNATService(ctx, h2,
+		// Support same non default security and transport options as
+		// original host.
+		libp2p.Security(libp2ptls.ID, libp2ptls.New),
+		libp2p.Security(secio.ID, secio.New),
+		libp2p.Transport(libp2pquic.NewTransport),
+		libp2p.DefaultTransports,
+	)
+
+	// The last steps to get fully up and running would be to
+	// connect to bootstrap peers (or any other peers) and bootstrap
+	// the DHT. We leave this commented as this is an example and the peer
+	// will die as soon as it finishes, so it is unnecessary to
+	// put strain on the network.
+
+	/*
+		// This connects to public bootstrappers
+		for _, addr := range dht.DefaultBootstrapPeers {
+			pi, _ := peer.AddrInfoFromP2pAddr(addr)
+			// We ignore errors as some bootstrap peers may be down
+			// and that is fine.
+			h2.Connect(ctx, *pi)
+		}
+
+		// Now that we have connected to some peers, we can bootstrap the DHT
+		// and discover even more peers.
+		network.  dht.Bootstrap(ctx)
```

Comment:
We can skip this now.

Reply:
ah. Is bootstrap automatic now from creation?

Reply:
Yes. We bootstrap as soon as we start and whenever we create a new connection when we have an empty routing table.